lib/filecoinpayment/utils.go

```diff
@@ -0,0 +1,146 @@
+	if err != nil {
+		return xerrors.Errorf("failed to get USDFC address: %w", err)
+	}
+
+	var railIds []*big.Int
+
+	current, err := ethClient.BlockNumber(ctx)
+	if err != nil {
+		return xerrors.Errorf("failed to get current block number: %w", err)
+	}
+
+	for _, payee := range payees {
+		rails, err := payment.GetRailsForPayeeAndToken(&bind.CallOpts{Context: ctx}, payee, tokenAddress, big.NewInt(0), big.NewInt(0))
+		if err != nil {
+			return xerrors.Errorf("failed to get Rails for: %w", err)
+		}
+
+		for _, r := range rails.Results {
+			if !r.IsTerminated {
+				railIds = append(railIds, r.RailId)
+			}
+
+			// If rail terminated in the last 30 days, we should consider it for settlement
+			if uint64(r.EndEpoch.Int64()) > current-(builtin.EpochsInDay*30) {
+				railIds = append(railIds, r.RailId)
+			}
+		}
+
+	}
+
+	var toSettle []*big.Int
+	bufferPeriod := big.NewInt(2 * builtin.EpochsInDay)
+	for _, rail := range railIds {
+		view, err := payment.GetRail(&bind.CallOpts{Context: ctx}, rail)
+		if err != nil {
+			return xerrors.Errorf("failed to get rail: %w", err)
+		}
+
+		// Let's only settle rail if it's our operator i.e., PDP related operator
+		if !lo.Contains(operators, view.Operator) {
+			continue
+		}
+
+		// If rail is terminated but not settled yet, settle it
+		if view.EndEpoch.Int64() > 0 {
+			if view.SettledUpTo.Cmp(view.EndEpoch) == -1 {
+				toSettle = append(toSettle, rail)
+			}
+		} else {
+			//If rail is not terminated, settle if SettledUpTo+LockupPeriod-2days > current block
+			threshold := big.NewInt(0).Add(view.SettledUpTo, view.LockupPeriod)
+			thresholdWithGrace := big.NewInt(0).Sub(threshold, bufferPeriod)
+
+			if thresholdWithGrace.Uint64() < current {
+				toSettle = append(toSettle, rail)
+			}
+		}
+	}
+
+	log.Debugw("Total number of rails to be settled", "total", len(railIds), "rails", railIds)
+
+	pabi, err := PaymentsMetaData.GetAbi()
+	if err != nil {
+		return xerrors.Errorf("failed to get Payments ABI: %w", err)
+	}
+
+	// Batch in the 10s to settle rail
+	calls := make([]multicall.Multicall3Call, 0, 10)
+	rails := make([]int64, 0, 10)
+	transactionsToSend := make(map[*types.Transaction][]int64)
+
+	for _, id := range toSettle {
+		data, err := pabi.Pack("settleRail", id, big.NewInt(int64(current)))
+		if err != nil {
+			return xerrors.Errorf("failed to pack data: %w", err)
+		}
+		calls = append(calls, multicall.Multicall3Call{
+			Target:   paymentContractAddr,
+			CallData: data,
+		})
+		rails = append(rails, id.Int64())
+		if len(calls) >= 10 {
+			tx, err := multicall.BatchCallGenerate(calls)
+			if err != nil {
+				return xerrors.Errorf("failed to generate batch call: %w", err)
+			}
+			transactionsToSend[tx] = rails
+			calls = make([]multicall.Multicall3Call, 0, 10)
+			rails = make([]int64, 0, 10)
+		}
+	}
+
+	for txToSend, railIDs := range transactionsToSend {
+		tx, err := sender.Send(ctx, from, txToSend, "settleRail")
+		if err != nil {
```

Comment:
first to fail will mean that all queued transactions fail, perhaps you want to do a per-transaction handling here - if transaction fails to send, don't record it in db, but continue to next queued transaction and deal with that separately

if there's something quirky about one particular rail, or for some reason we bust gas limit with the batch-of-10, then it could get stuck in our throat and stop us digesting any settlements, which will suck more than just a few failing to settle.